packages/embedder/mergeAsyncIterators.ts

```diff
@@ -15,82 +15,79 @@ type Result<T extends AsyncIterator<any>> = UnYield<Awaited<ReturnType<T['next']
       }
+    } catch (error) {
+      resolveResult(() => {
+        throw error
+      })
     }
-    const resolveResult = (resultThunk: ResultThunk) => {
-      if (capability === undefined) {
-        queuedResults.push(resultThunk)
-      } else {
-        capability.resolve(resultThunk)
-      }
+  }
+  const resolveResult = (resultThunk: ResultThunk) => {
+    if (capability === undefined) {
+      queuedResults.push(resultThunk)
+    } else {
+      capability.resolve(resultThunk)
     }
+  }
 
-    try {
-      // Begin all iterators
-      for (const [idx, iterable] of iterators.entries()) {
-        void getNext(idx, iterable)
+  // Begin all iterators
+  for (const [idx, iterable] of iterators.entries()) {
+    void getNext(idx, iterable)
+  }
+
+  const it = {
+    [Symbol.asyncIterator]: () => it,
+    next: async () => {
+      const nextQueuedResult = queuedResults.shift()
+      if (nextQueuedResult !== undefined) {
+        return {done: false as const, value: nextQueuedResult()}
+      }
+      if (count === 0) {
+        return {done: true as const, value: undefined}
       }
 
-      // Delegate to iterables as results complete
-      while (true) {
-        while (true) {
-          const nextQueuedResult = queuedResults.shift()
-          if (nextQueuedResult === undefined) {
-            break
-          } else {
-            yield nextQueuedResult()
-          }
-        }
-        if (count === 0) {
-          break
-        } else {
-          // Promise.withResolvers() is not yet implemented in node
-          capability = {
-            resolve: undefined as any,
-            reject: undefined as any,
-            promise: undefined as any
-          }
-          capability.promise = new Promise((res, rej) => {
-            capability!.resolve = res
-            capability!.reject = rej
-          })
-          const nextResult = await capability.promise
-          if (nextResult === null) {
-            break
-          } else {
-            capability = undefined
-            yield nextResult()
-          }
-        }
+      // Promise.withResolvers() is not yet implemented in node
+      capability = {
+        resolve: undefined as any,
+        reject: undefined as any,
+        promise: undefined as any
+      }
+      capability.promise = new Promise((res, rej) => {
+        capability!.resolve = res
+        capability!.reject = rej
+      })
+      const nextResult = await capability.promise
+      if (nextResult === null) {
+        return {done: true as const, value: undefined}
+      } else {
+        capability = undefined
+        return {done: false as const, value: nextResult()}
       }
-    } catch (err) {
-      // Unwind remaining iterators on failure
-      try {
-        await Promise.all(iterators.map((iterator) => iterator.return?.()))
-      } catch {}
-      throw err
+    },
+    return: async () => {
+      await Promise.allSettled(iterators.map((iterator) => iterator.return?.()))
```

Comment:
This refactoring was done so we can define this return function. This is necessary so all streams are notified of the return. The async generator function we had previously would hang in `await capability.promise` and handle the return only afterwards. If all embedders are idle, the promise would never resolve.

Reply:
interesting! really good find having to define our own `return` here